colorizer_data/converter.py

```diff
@@ -176,12 +178,89 @@ def _write_data(
     )
 
 
+def _get_raw_backdrop_paths(
+    grouped_frames: DataFrameGroupBy, column_name: str
+) -> List[str | None]:
+    # Get the backdrop paths for each frame.
+    backdrop_paths = []
+    for _group_name, frame in grouped_frames:
+        row = frame.iloc[0]
+        if column_name not in row or row[column_name] is None:
+            backdrop_paths.append(None)
+        else:
+            backdrop_paths.append(row[column_name])
+    return backdrop_paths
+
+
+def _write_backdrop_from_column(
+    backdrop_column: str,
+    grouped_frames: DataFrameGroupBy,
+    writer: ColorizerDatasetWriter,
+    config: ConverterConfig,
+):
+    backdrop_metadata = BackdropMetadata(
+        frames=_get_raw_backdrop_paths(grouped_frames, backdrop_column),
+        name=backdrop_column,
+        key=sanitize_key_name(backdrop_column),
+    )
+
+    # Override metadata if provided
+    if config.backdrop_info is not None and backdrop_column in config.backdrop_info:
+        override_metadata = config.backdrop_info.get(backdrop_column)
+        backdrop_metadata = merge_dictionaries(backdrop_metadata, override_metadata)
+        # Just to be safe, sanitize the key
+        backdrop_metadata["key"] = sanitize_key_name(backdrop_metadata["key"])
+
+    # Iterate over all the backdrop paths and rewrite them to be relative to
+    # the dataset directory, copying the file into the dataset dir if necessary.
+    updated_frame_paths = []
+    for frame_number, raw_backdrop_image_path in enumerate(backdrop_metadata["frames"]):
+        if raw_backdrop_image_path is None:
+            updated_frame_paths.append(None)
+            continue
+        backdrop_image_path = pathlib.Path(
+            sanitize_path_by_platform(raw_backdrop_image_path)
+        )
+        if not os.path.exists(backdrop_image_path):
+            raise FileNotFoundError(
+                f"Backdrop image '{backdrop_image_path}' does not exist. Please check the path and try again."
+            )
+        if writer.outpath in backdrop_image_path.parents:
+            # Path exists and is already in the dataset directory.
+            relative_path = backdrop_image_path.relative_to(writer.outpath)
+            updated_frame_paths.append(relative_path.as_posix())
+        else:
+            # Path exists but is not in the dataset directory. Copy it.
+            folder = writer.outpath / backdrop_metadata["key"]
+            folder.mkdir(parents=True, exist_ok=True)
+            dst_path = folder / f"image_{frame_number}.png"
+            shutil.copy(backdrop_image_path, dst_path)
+            relative_path = dst_path.relative_to(writer.outpath)
+            updated_frame_paths.append(relative_path.as_posix())
```

Comment:
This could maybe be even smarter and check if the two files are the same before copying?

Reply:
Also, should we use `shutil.copy2` to preserve the creation + modification time of images? (I'd have to edit some of my unit tests if so)
https://docs.python.org/3/library/shutil.html

Reply:
1. It's probably okay unless the file I/O starts slowing this script down significantly...
2. `shutil.copy2` preserves the creation and last modified time of a file, copying it without changing either of those properties. It could be useful when using a file explorer to check the creation time of the original image, but we don't make any promises about the file copying/creation behavior in the docstring spec right now.

Reply:
I still think the image file transfer part of backdrops can be all-or-nothing, kinda like the other image processing steps.  It basically goes without saying that copying several hundred images will be a lot slower than just writing out the json data.

Sounds like there is no current use case for using `copy2` in this codebase, and in fact it could be deceptive in hiding the true last modified time?

Reply:
Okay, I tried working on a smart copy system but I need to sit down with it for a bit! It'll be a follow-up PR, I think.